datafusion/expr/src/logical_plan/ddl.rs

```diff
@@ -288,8 +290,234 @@ impl PartialOrd for CreateExternalTable {
     }
 }
 
+impl CreateExternalTable {
+    pub fn try_new(fields: CreateExternalTableFields) -> Result<Self> {
+        let CreateExternalTableFields {
```

Comment:
@jonahgao  I wonder if you have some time to offer an opinion on this pattern (a way to allow pattern matching but still ensure a `struct` has to be valudated as part of construction

Reply:
I think `CreateExternalTableFields` makes things too complex. It has exactly the same fields as `CreateExternalTable`, and users will feel confused and need to spend time distinguishing and using them.

In my opinion, it is enough to perform all the validity checks in `CreateExternalTableBuilder::build`, such as missing required fields and duplicate column names. `CreateExternalTable`s typically come from the SQL planner of DataFusion, where we have already used the builder. For other scenarios, we can encourage users in the documentation to use the Builder API to create `CreateExternalTable`s .

Reply:
If we don't block public construction, the downstream projects won't know to migrate to the builder API.

Reply:
The verification is in the `try_new()` method used by the builder. If the builder is not used, and this new `try_new` method isn't used directly either, the verification won't run.
I believe the downside of not migrating may be no error message at all.